### ISSUE TITLE
Numerical instabilities for gradient and hessian

### ISSUE THREAD

**Issue by sejlermonster** [dmlc/xgboost#2955]
Operating System:
Windows 7 enterprise

Package used (python/R/jvm/C++):
Python

xgboost version used:
0.6

The python version and distribution:
Python 2.7.13, Anaconda custom 64-bit <default, May 11 2017>

Python packages:
('NumPy', '1.12.1')
('SciPy', '0.19.0')
('Scikit-Learn', '0.19.0')
## 

I have been looking into the stability of the binary log loss computation of XGBoost. When looking at the implementation I'm not sure how it can be stable. 

predt is computed with the Sigmoid function, which does not seem to ensure no overflow or underflow. The  FirstOrderGradient function does not ensure that any round-off or similar could not lead to negative values. 
The SecondOrderGradient Function does use an epsilon value of 1e-16f to ensure no value lower than this. 

```
struct LogisticRegression {
  static bst_float PredTransform(bst_float x) { return common::Sigmoid(x); }
  static bool CheckLabel(bst_float x) { return x >= 0.0f && x <= 1.0f; }
  static bst_float FirstOrderGradient(bst_float predt, bst_float label) { return predt - label; }
  static bst_float SecondOrderGradient(bst_float predt, bst_float label) {
    const float eps = 1e-16f;
    return std::max(predt * (1.0f - predt), eps);
  }
  static bst_float ProbToMargin(bst_float base_score) {
    CHECK(base_score > 0.0f && base_score < 1.0f)
        << "base_score must be in (0,1) for logistic loss";
    return -std::log(1.0f / base_score - 1.0f);
  }
  static const char* LabelErrorMsg() {
    return "label must be in [0,1] for logistic regression";
  }
  static const char* DefaultEvalMetric() { return "rmse"; }
};
// logistic loss for binary classification task.
struct LogisticClassification : public LogisticRegression {
  static const char* DefaultEvalMetric() { return "error"; }
};
```
To try to practically test to this i grew very deep trees
I got a stackoverflow when printing the tree, not when actually fitting the tree, so I'm not really sure if these are related.

First example with lossguided growth policy. I am aware that this overfits like hell but it was just to investigate the code above.
```
X, y = make_classification(n_samples=9999, n_features=8, n_informative=4, n_classes=2, n_clusters_per_class=1, random_state=0)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=7)

eval_set = [(X_train, y_train), (X_test, y_test)]
model = XGBClassifier(max_depth=0, min_child_weight=0, grow_policy='lossguide')
model.fit(X_train, y_train, eval_metric=["error", "logloss"], eval_set=eval_set)
print model.feature_importances_
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: %.2f%%" % (accuracy * 100.0))
```

Similar code example where I am training and testing on the same data. Despite that it should overfit to this data, it has a random guessing accuracy of approximately 50%. 
```
X, y = make_classification(n_samples=9999, n_features=8, n_informative=4, n_classes=2, n_clusters_per_class=1, random_state=0)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=7)

lgbmClassi = lgb.LGBMClassifier(min_child_weight=0, max_depth=-1)
lgbmClassi.fit(X_train, y_train)
print lgbmClassi.feature_importances_
y_pred = lgbmClassi.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: %.2f%%" % (accuracy * 100.0))

print "xgboost started...."
eval_set = [(X_train, y_train), (X_train, y_train)]
model = XGBClassifier(max_depth=5000, min_child_weight=0)
model.fit(X_train, y_train, eval_metric=["error", "logloss"], eval_set=eval_set)
print model.feature_importances_
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: %.2f%%" % (accuracy * 100.0))
```

Are my thoughts on XGBoosts computation of gradient and hessian for binary log loss right? Then would it not be smarter to add an epsilon to the computation of sigmoid and gradient? 
Is this computation of gradient and hessian even related to the problem I had when growing trees very deep with XGBoost? 
I hope this is not way of, if so I hope someone can explain why this is not a problem.

**Comment by khotilov** [comment]
No, you don't want non-zero gradient when there's no gradient. But it's a different story for the hessian.

To use the loss-guided algorithm, you have to choose `tree_method='hist'`, e.g.,
```
model = XGBClassifier(max_depth=0, grow_policy='lossguide', max_leaves=32, tree_method='hist')
```

With your second example, I see the overfitting, as expected (with the recent development code). The crash during get_dump() is interesting... However, I would advise against using such a large tree depth, as the tree nodes are enumerated using 32 bit integers (https://github.com/dmlc/xgboost/issues/2788#issuecomment-337762210)

**Comment by sejlermonster** [comment]
You are right, if there is no gradient, you don't want a non-zero gradient. But with very large x values in the sigmoid function can result in round-off error? Can this not give unwanted effects on the gradient and hessian computation? But I guess its unavoidable and won't really have any practical effect
```
inline float Sigmoid(float x) {
  return 1.0f / (1.0f + std::exp(-x));
}
```
Is the hessian value set to epsilon to avoid divison by zero if XGBoost uses Newton's method? Or can you elaborate on that?

For the second example, the crash does not seem to be related to the limit of a 32 bit integer, does it?  As you say, XGBoost is able to lower the log loss in each iteration. The stacoverflow actually only happens if you print the feature importance of the model.

Of cause one could generally advice against growing this deep, but other regularization measurements could be chosen and large data sets could also grow deep trees without introducing overfitting.
